cartographer_ros/cartographer_ros/offline_node.cc

```diff
@@ -132,101 +153,139 @@ void RunOfflineNode(const MapBuilderFactory& map_builder_factory) {
       },
       false /* oneshot */, false /* autostart */);
 
-  for (const std::string& bag_filename : bag_filenames) {
+  std::vector<std::string> specified_sensor_topics =
+      cartographer_ros::SplitString(FLAGS_sensor_topics, ',');
+
+  if (specified_sensor_topics.empty()) {
+    specified_sensor_topics =
+        std::vector<std::string>(bag_filenames.size(), std::string());
+  } else if (specified_sensor_topics.size() == 1) {
+    specified_sensor_topics.insert(specified_sensor_topics.end(),
+                                   bag_filenames.size() - 1,
+                                   specified_sensor_topics.at(0));
+  }
+  CHECK_EQ(specified_sensor_topics.size(), bag_filenames.size());
+
+  std::unordered_map<int, std::unordered_set<std::string>>
+      expected_sensor_topics;
+  std::unordered_map<int, int> bag_id_to_trajectory_id;
+  PlayableBagMultiplexer playable_bag_multiplexer;
+  for (size_t current_bag_index = 0; current_bag_index < bag_filenames.size();
+       ++current_bag_index) {
+    const std::string& bag_filename = bag_filenames.at(current_bag_index);
     if (!::ros::ok()) {
-      break;
+      return;
     }
-
-    const int trajectory_id =
-        node.AddOfflineTrajectory(expected_sensor_ids, trajectory_options);
-
-    rosbag::Bag bag;
-    bag.open(bag_filename, rosbag::bagmode::Read);
-    rosbag::View view(bag);
-    const ::ros::Time begin_time = view.getBeginTime();
-    const double duration_in_seconds = (view.getEndTime() - begin_time).toSec();
-
-    // We need to keep 'tf_buffer' small because it becomes very inefficient
-    // otherwise. We make sure that tf_messages are published before any data
-    // messages, so that tf lookups always work.
-    std::deque<rosbag::MessageInstance> delayed_messages;
-    // We publish tf messages one second earlier than other messages. Under
-    // the assumption of higher frequency tf this should ensure that tf can
-    // always interpolate.
-    const ::ros::Duration kDelay(1.);
-    for (const rosbag::MessageInstance& msg : view) {
-      if (!::ros::ok()) {
-        break;
+    std::unordered_set<std::string> current_expected_sensor_topics;
+    if (specified_sensor_topics.at(current_bag_index).empty()) {
+      for (const std::string& topic : node.ComputeDefaultTopics(
+               trajectory_options.at(current_bag_index))) {
+        CHECK(current_expected_sensor_topics
```

Comment:
move out of if? it's the same in both cases
(create list of topics before and use that)